src/lava/lib/optimization/solvers/generic/scif/models.py

```diff
@@ -16,145 +17,176 @@
-        state_to_intg[wta_spk_idx] = self.neg_tau_ref  # Post spk refractory
-        spk_hist_to_intg[wta_spk_idx] |= 1
+        state_to_intg = state_to_intg + lfsr_to_intg + cnstr_to_intg + \
+            step_size_to_intg
+        np.clip(state_to_intg, a_min=0, a_max=2 ** 24 - 1,
+                out=state_to_intg)
 
         # Assign all temporary states to state Vars
         self.state[intg_idx] = state_to_intg
-        self.cnstr_intg[intg_idx] = cnstr_to_intg
-        self.spk_hist[intg_idx] = spk_hist_to_intg
 
-        return wta_spk_idx
+        # WTA spike indices when threshold is exceeded
+        wta_enter_on_state = np.where(self.state >= self.theta)
 
-    def _refractory_dynamics(self, rfct_idx):
-        # Split/fork state variables u, v, beta
+        # Spiking neuron voltages go in refractory (sustained_on_tau < 0)
+        self.state[wta_enter_on_state] = self.sustained_on_tau
+        self.spk_hist[wta_enter_on_state] |= 1
+
+        return wta_enter_on_state
+
+    def _refractory_dynamics(self, rfct_idx, local_conflict):
+        # Split/fork state variables
         state_in_rfct = self.state[rfct_idx]  # voltages in refractory
-        cnstr_in_rfct = self.cnstr_intg[rfct_idx]  # currents in refractory
-        spk_hist_in_rfct = self.spk_hist[rfct_idx]  # beta in refractory
+        spk_hist_in_rfct = self.spk_hist[rfct_idx]
 
         # Refractory dynamics
         state_in_rfct += 1  # voltage increments by 1 every step
         spk_hist_in_rfct |= 3
-        # Second set of unsatisfied WTA indices based on v and u in refractory
-        wta_unsat_idx_2 = \
-            np.where(np.logical_or(state_in_rfct == 0, cnstr_in_rfct < 0))
 
-        # Reset voltage of unsatisfied WTA in refractory
-        state_in_rfct[wta_unsat_idx_2] = 0
-        spk_hist_in_rfct[wta_unsat_idx_2] &= 2
+        # The following proxy avoids picking up states that have gone to zero
+        # through their "natural" dynamics, and only picks up indices of
+        # those states that have become zero by coming out of sustained ON
+        # period.
+        state_proxy_for_rfct = np.ones_like(self.state)
+        state_proxy_for_rfct[rfct_idx] = state_in_rfct
+        spk_hist_proxy = np.zeros_like(self.spk_hist)
+        spk_hist_proxy[rfct_idx] = spk_hist_in_rfct
+
+        # Second set of unsatisfied WTA indices based on refractory
+        wta_enter_off_state = np.where(
+            np.logical_or(state_proxy_for_rfct == 0, np.logical_and(
+                spk_hist_proxy & 1, local_conflict)))
 
         # Assign all temporary states to state Vars
         self.state[rfct_idx] = state_in_rfct
-        self.cnstr_intg[rfct_idx] = cnstr_in_rfct
         self.spk_hist[rfct_idx] = spk_hist_in_rfct
 
-        return wta_unsat_idx_2
+        # Reset voltage of unsatisfied WTA in refractory
+        self.state[wta_enter_off_state] = 0
+        self.spk_hist[wta_enter_off_state] &= 0xFE  # Last two bits &= 10
```

Comment:
This does not implement &= 10

0xFE = 0b11111110

Please replace by desired binary number.